# Предсказание уровня дохода

In [161]:
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [162]:
data = pd.read_csv('adult.csv')
data.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [163]:
selected_columns = data[ ['education', 'capital-gain', 'capital-loss'] ]

In [164]:
X = pd.get_dummies(selected_columns, columns = ['education'])
X.head()

capital-gain  capital-loss  education_10th  education_11th  education_12th  \
0             0             0               0               1               0   
1             0             0               0               0               0   
2             0             0               0               0               0   
3          7688             0               0               0               0   
4             0             0               0               0               0   

   education_1st-4th  education_5th-6th  education_7th-8th  education_9th  \
0                  0                  0                  0              0   
1                  0                  0                  0              0   
2                  0                  0                  0              0   
3                  0                  0                  0              0   
4                  0                  0                  0              0   

   education_Assoc-acdm  education_Assoc-voc  education_Bachelors  \
0                     0                    0                    0   
1                     0                    0                    0   
2                     1                    0                    0   
3                     0                    0                    0   
4                     0                    0                    0   

   education_Doctorate  education_HS-grad  education_Masters  \
0                    0                  0                  0   
1                    0                  1                  0   
2                    0                  0                  0   
3                    0                  0                  0   
4                    0                  0                  0   

   education_Preschool  education_Prof-school  education_Some-college  
0                    0                      0                       0  
1                    0                      0                       0  
2                    0                      0                       0  
3                    0                      0                       1  
4                    0                      0                       1

In [165]:
target = data['income']
target.head()

0    <=50K
1    <=50K
2     >50K
3     >50K
4    <=50K
Name: income, dtype: object

In [166]:
le = LabelEncoder()
le.fit( target )
Y = pd.Series( le.transform( data['income'] ) )
Y.head()

0    0
1    0
2    1
3    1
4    0
dtype: int64

In [167]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.9, random_state = 0)

In [168]:
model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [169]:
model.score(X_train, Y_train)

0.8192055692055692

In [170]:
model.score(X_test, Y_test)

0.806087629100505

In [171]:
for i, C in enumerate((100, 1, 0.01)):
    print( i, C )

0 100
1 1
2 0.01


In [172]:
results = []
penalties = ['l1', 'l2']
for i, C in enumerate((10000, 1000, 100, 10, 1, 0.1, 0.01, 0.001, 0.0001)):
    print('C = {}'.format(C))
    for penalty in penalties:
        print(penalty)
        model = LogisticRegression(C=C, penalty=penalty)

        model.fit(X_train, Y_train)
        
        results.append({'C': C, 'penalty': penalty, 'score': 100*(model.score(X_test, Y_test))})

C = 10000
l1
l2
C = 1000
l1
l2
C = 100
l1
l2
C = 10
l1
l2
C = 1
l1
l2
C = 0.1
l1
l2
C = 0.01
l1
l2
C = 0.001
l1
l2
C = 0.0001
l1
l2


In [173]:
results

[{'C': 10000, 'penalty': 'l1', 'score': 80.85217707812002},
 {'C': 10000, 'penalty': 'l2', 'score': 80.8885754583921},
 {'C': 1000, 'penalty': 'l1', 'score': 80.86810136948905},
 {'C': 1000, 'penalty': 'l2', 'score': 80.8885754583921},
 {'C': 100, 'penalty': 'l1', 'score': 80.85217707812002},
 {'C': 100, 'penalty': 'l2', 'score': 80.8885754583921},
 {'C': 10, 'penalty': 'l1', 'score': 80.85445197688703},
 {'C': 10, 'penalty': 'l2', 'score': 80.85217707812002},
 {'C': 1, 'penalty': 'l1', 'score': 81.1706629055007},
 {'C': 1, 'penalty': 'l2', 'score': 80.6087629100505},
 {'C': 0.1, 'penalty': 'l1', 'score': 80.91814914236316},
 {'C': 0.1, 'penalty': 'l2', 'score': 80.54734064334137},
 {'C': 0.01, 'penalty': 'l1', 'score': 80.843077483052},
 {'C': 0.01, 'penalty': 'l2', 'score': 80.59283861868147},
 {'C': 0.001, 'penalty': 'l1', 'score': 78.44078438509486},
 {'C': 0.001, 'penalty': 'l2', 'score': 80.18335684062059},
 {'C': 0.0001, 'penalty': 'l1', 'score': 77.93803175758678},
 {'C': 0.000

In [174]:
scores = []
for dict in results:
    scores.append(dict['score'])

### Наилучший результат с этой моделью без кросс-валидации

In [175]:
results[scores.index(max(scores))]

{'C': 1, 'penalty': 'l1', 'score': 81.1706629055007}

In [176]:
model = LogisticRegression(C=1, penalty='l1')
model.fit(X_train, Y_train)
predictions = model_log.predict(X_test)
roc_auc_score( Y_test, predictions )

0.6654471573036546

In [177]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import GridSearchCV

In [178]:
penalties = ['l1', 'l2']
Cs = [10000, 1000, 100, 10, 1, 0.1, 0.01, 0.001, 0.0001]

In [179]:
def get_best_result_log_reg(Cs, penalties):
    results_cross_val = [] 
    results_grid_search = []
    
    #cross_val
    for C in Cs:
        for penalty in penalties:
            for cv in range(2, 10): 
                score = cross_val_score( LogisticRegression(C=C, penalty=penalty), X, Y, cv = cv, scoring = 'roc_auc' )
                results_cross_val.append({'C': C, 'penalty': penalty, 'score': 100*np.mean(score)})   
    
    #grid_search
    params = {'C': Cs, 'penalty': penalties}
    for cv in range(2, 10):    
        gs = GridSearchCV(LogisticRegression(), params, cv = cv, scoring = 'roc_auc')
        gs.fit( X, Y )
        results_grid_search.append({'C': best_params['C'], 'penalty': best_params['penalty'], 'score': 100*(gs.best_score_)})  


    scores_grid_search = []
    for dict in results_grid_search:
        scores_grid_search.append(dict['score'])
        
    scores_cross_val = []
    for dict in results_cross_val:
        scores_cross_val.append(dict['score'])
        
    if max(scores_grid_search) > max(scores_cross_val):
        print(results_grid_search[scores_grid_search.index(max(scores_grid_search))])
    else:
        print(results_cross_val[scores_cross_val.index(max(scores_cross_val))])


In [180]:
get_best_result_log_reg(Cs, penalties)

{'C': 10000, 'penalty': 'l1', 'score': 76.91365803295633}


In [181]:
model_log = LogisticRegression(C=10000, penalty='l1')

In [182]:
model_log.fit(X_train, Y_train)

LogisticRegression(C=10000, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

### Наилучший результат с этой моделью c кросс-валидацией, где классификатор - логистическая регрессия

In [183]:
model_log.score(X_test, Y_test)

0.8086810136948905

In [184]:
predictions_log = model_log.predict(X_test)

In [185]:
roc_auc_score( Y_test, predictions_log )

0.6657307647908923

In [130]:
from sklearn.tree import DecisionTreeClassifier

In [147]:
def get_best_result_decision_tree():
    results_cross_val = [] 
    results_grid_search = []
    
    for depth in range (1, 26):
        for cv in range(2, 10): 
            #cross_val
            score = cross_val_score( DecisionTreeClassifier( max_depth = depth ), X, Y, cv = cv, scoring = 'roc_auc')
            results_cross_val.append({'Depth': depth, 'cv': cv, 'score': 100*np.mean(score)})
            
    #grid_search        
    for cv in range(2, 10): 
        params = [ {'max_depth': list( range(1, 26) )} ]
        gs = GridSearchCV(DecisionTreeClassifier(), param_grid = params, cv = cv, scoring = 'roc_auc')
        gs.fit( X, Y )
        results_grid_search.append({'Depth': depth, 'cv': cv, 'score': 100*(gs.best_score_)})  


    scores_grid_search = []
    for dict in results_grid_search:
        scores_grid_search.append(dict['score'])
        
    scores_cross_val = []
    for dict in results_cross_val:
        scores_cross_val.append(dict['score'])
        
    if max(scores_grid_search) > max(scores_cross_val):
        print(results_grid_search[scores_grid_search.index(max(scores_grid_search))])
    else:
        print(results_cross_val[scores_cross_val.index(max(scores_cross_val))])

In [148]:
get_best_result_decision_tree()

{'Depth': 25, 'cv': 4, 'score': 80.48715789485594}


### Наилучший результат с этой моделью c кросс-валидацией, где классификатор - дерево решений
#### Самый наилучший результат. Стартовая модель, данная на уроке - start-model, модель с кросс-валидацией(Decision Tree) - cross-model
#### start-model.score = 81.17, start-model.roc_auc_score = 66.5
#### cross-model.score = 82.797, cross-model.roc_auc_score = 69.25

In [155]:
model_dec_tree = DecisionTreeClassifier( max_depth = 25 )
model_dec_tree.fit(X_train, Y_train)
model_dec_tree.score(X_test, Y_test)

0.8279039082760817

In [186]:
predictions_dec_tree = model_dec_tree.predict(X_test)

In [187]:
roc_auc_score( Y_test, predictions_dec_tree )

0.6925951540786744